File: 300. Longest Increasing Subsequence.py
# -*- coding: utf-8 -*-
"""
Created on Tue Aug 21 20:27:42 2018

@author: TZLMYQ
"""

class Solution(object):
    def lengthOfLIS(self, nums):
        tails = [0] * len(nums)
        size = 0
        for x in nums:
            i, j = 0, size
            while i != j:
                m = (i + j) / 2
                if tails[m] < x:
                    i = m + 1
                else:
                    j = m
            tails[i] = x
            size = max(i + 1, size)
        return size
    
solution = Solution()
nums = [-10, -3, -2, -9] 
print solution.lengthOfLIS(nums)

File: DP/300. Longest Increasing Subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int> dp;
        for(auto& num:nums)
        {
            if(dp.size() == 0)
            {
                dp.push_back(num);
            }
            else if(num > dp.back())
                dp.push_back(num);
            else if(num < dp[0])
                dp[0] = num;
            else 
            {
                int i = dp.size() - 1;
                while(i > 0)
                {
                    if(num > dp[i - 1])
                    {
                        dp[i] = num;
                        break;
                    }
                    i--;
                }                
            }
        }
        return dp.size();
    }
};
